stanford/bin/audit-iam-users

```diff
@@ -0,0 +1,163 @@
+    print('### TODO: delete password')
+    users = get_users_with_unused_passwords(all_users)
+    print_items(users)
+    print('')
+
+
+def audit_users_cant_log_in(all_users):
+    print('## inaccessible users')
+    print('### TODO: delete user')
+    users = get_users_cant_log_in(all_users)
+    print_items(users)
+    print('')
+
+
+def audit_users_never_logged_in(all_users):
+    print('## unused users')
+    print('### TODO: delete password and deactivate key')
+    users = get_users_never_logged_in(all_users)
+    print_items(users)
+    print('')
+
+
+def get_user_data():
+    client = boto3.client('iam')
+    response = client.generate_credential_report()
+    if response.get('State') != 'COMPLETE':
+        print('response.State', response.get('State'))
+        time.sleep(60)
+    time.sleep(5)
+    response = client.get_credential_report()
+    content = response['Content']
+    lines = content.splitlines()
+    data = csv.DictReader(lines)
+    return data
+
+
+def get_users_cant_log_in(data):
+    for user in data:
+        if is_password_enabled(user):
+            continue
+        if are_keys_active(user):
+            continue
+        yield user
+
+
+def get_users_never_logged_in(users):
+    for user in users:
+        if has_used_password(user):
+            continue
+        if has_used_key(user):
+            continue
+        yield user
+
+
+def get_users_with_unused_keys(users):
+    for user in users:
+        if has_unused_key(user):
+            yield user
+
+
+def get_users_with_unused_passwords(all_users):
+    for user in all_users:
+        if has_unused_password(user):
+            yield user
+
+
+def has_active_key(user):
+    if user['access_key_1_active']:
+        return True
+    if user['access_key_2_active']:
+        return True
+    return False
+
+
+def has_unused_key(user):
+    if user['access_key_1_active'] == 'true':
+        if user['access_key_1_last_used_date'] == 'N/A':
+            return True
+    if user['access_key_2_active'] == 'true':
+        if user['access_key_2_last_used_date'] == 'N/A':
+            return True
+    return False
+
+
+def has_unused_password(user):
+    if user['password_enabled'] == 'true':
+        if user['password_last_used'] == 'no_information':
+            return True
+    return False
+
+
+def has_used_key(user):
+    if user['access_key_1_last_used_date'] != 'N/A':
+        return True
+    if user['access_key_2_last_used_date'] != 'N/A':
+        return True
+    return False
+
+
+def has_used_password(user):
+    if user['password_enabled'] == 'true':
+        if user['password_last_used'] == 'no_information':
+            return False
+        return True
+    return False
+
+
+def is_password_enabled(user):
+    is_enabled = user['password_enabled'] == 'true'
+    return is_enabled
+
+
+def print_items(users):
+    for user in users:
+        print('- ' + user['user'])
+
+
+def main():
+    print('# User credential report')
+    print('')
```

Comment:
Is there really no more elegant way to add blank lines?

Reply:
Nope. Though I don't find this particularly inelegant;
it's printing nothing.